### PR TITLE
Remove `mut` self reference from LogExporter::export() method.

### PR DESCRIPTION
**Pull request by lalitb**
## Changes

As discussed here - https://github.com/open-telemetry/opentelemetry-rust/pull/2374#discussion_r1866729502
The changes are minimal, and the stress test for the logs now includes an exporter call over the future, causing a slight regression.

## Merge requirement checklist
* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
I think this is the right direction. Our processors are only requiring `self` and not `but self`, so there seems to be no strong reason for exporter methods to require `mut self`. Also looked at `tracing-subscriber` - its on_event method is also [just requiring `self` ](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-appender-tracing/src/layer.rs#L157), and it lets layers manage its state via interior mutability.

Requiring `mut self` would prevent exporters's export() method from being called in parallel. While this is a spec requirement now, we expect to allow exporters to be called parallel in certain situations. Instead of adding another trait for that, it seems okay to use `self` itself and let exporters handle state via interior mutability, if they need to.

Would love to get more feedback on this, as we are close to declaring stability, so can't make breaking changes after that.

**Comment by lalitb** [comment]
agree, would prefer to have closer on this sooner, the slight regression seen in stress test can then be evaluated to improve with #2374.

**Comment by scottgerring** [comment]
Cool! LGTM. As a ordinary user of the library I'd prefer for any "incidental mutability" of a particular exporter to be handled internally, because it makes _my_ life easier. I reckon as users of the library implementing _custom_ exporters will be a much smaller set of folks, it makes sense for the ergonomics to focus on improving the lives of the former group.

A couple of minor observations that don't think really affect the PR.

It's a bit unfortunate that [tonic only provides](https://docs.rs/tonic/latest/tonic/service/interceptor/trait.Interceptor.html) a `(&mut self...)` for the interceptor trait call. As far as I can tell the only place we're using the interceptor is in `add_metadata`, and it's definitely not mutating anything on itself. The lock is necessary to keep the borrow-checker happy, but only because of this.

I see `PushMetricExporter` doesn't have a mutable export fn ✅ - but `SpanExporter` still does - is the story there different, or should this drop the `&mut self` too?

https://github.com/open-telemetry/opentelemetry-rust/blob/957659fac9217534e64d36f80bddc8360345fde5/opentelemetry-sdk/src/export/trace.rs#L20-L33

**Comment by lalitb** [comment]
> It's a bit unfortunate that [tonic only provides](https://docs.rs/tonic/latest/tonic/service/interceptor/trait.Interceptor.html) a (&mut self...) for the interceptor trait call. As far as I can tell the only place we're using the interceptor is in add_metadata, and it's definitely not mutating anything on itself. The lock is necessary to keep the borrow-checker happy, but only because of this.

Thanks @scottgerring for the review. That's the only unfortunate aspect of this change, which we discussed in the last community meeting - exporters need to acquire an uncontested lock for their internal mutable state during each export to satisfy the borrow checker. That said, we agreed the overhead should be minimal, likely no more than an atomic operation.

**Comment by cijothomas** [comment]
> > It's a bit unfortunate that [tonic only provides](https://docs.rs/tonic/latest/tonic/service/interceptor/trait.Interceptor.html) a (&mut self...) for the interceptor trait call. As far as I can tell the only place we're using the interceptor is in add_metadata, and it's definitely not mutating anything on itself. The lock is necessary to keep the borrow-checker happy, but only because of this.
> 
> Thanks @scottgerring for the review. That's the only unfortunate aspect of this change, which we discussed in the last community meeting - exporters need to acquire an uncontested lock for their internal mutable state during each export to satisfy the borrow checker. That said, we agreed the overhead should be minimal, likely no more than an atomic operation.

Can exporters leverage other interior mutability techniques (`RefCell`) other than mutex locks?

**Comment by TommyCpp** [comment]
> Yes, I believe we should do the similar change for SpanExporter too.

+1 on keeping then consistent

**Review by TommyCpp — APPROVED** [review]
LGTM. We have to handle mutability somewhere along the call stack. Having it in the exporter seems fine.

**Comment by lalitb** [comment]
> Can exporters leverage other interior mutability techniques (RefCell) other than mutex locks?

Yes, I believe exporters could use either mutexes or RefCell here for interior mutability, with similar overhead. Uncontested mutexes will involve atomic operations, while RefCell relies on lightweight runtime borrowing checks.

File: stress/src/logs.rs [hunk 1]
```diff
@@ -6,26 +6,38 @@
...
+#[async_trait]
+impl LogExporter for NoopExporter {
+    async fn export(&self, _: LogBatch<'_>) -> LogResult<()> {
+        LogResult::Ok(())
+    }
+}
 
 #[derive(Debug)]
-pub struct NoOpLogProcessor;
+pub struct NoOpLogProcessor {
```
[the remark this answers is not in the document]
Comment (lalitb) [reply stress/src/logs.rs]: yes, changed to Mock now.

**Comment by scottgerring** [comment]
The only reason it needs the mutex at the moment is that trait we're using to add metadata has the mut on it. I wonder if there's a different way to put the metadata on that doesn't use that trait?

To be clear - I think it's important to get this in to get the interface locked down, so i'm not suggesting this as part of this change - just thinking out loud!

**Review by cijothomas — APPROVED** [review]
LGTM.

**Review by Kc1296 — APPROVED** [review]

**Pull request merged**